Flow: add Elasticsearch exporter component (#4219)
grafana/agent: docs/sources/flow/reference/components/prometheus.exporter.elasticsearch.md

---
# NOTE(rfratto): the title below has zero-width spaces injected into it to
# prevent it from overflowing the sidebar on the rendered site. Be careful when
# modifying this section to retain the spaces.
#
# Ideally, in the future, we can fix the overflow issue with css rather than
# injecting special characters.

title: prometheus.exporter.elasticsearch
---

# prometheus.exporter.elasticsearch
The `prometheus.exporter.elasticsearch` component embeds
[elasticsearch_exporter](https://github.com/prometheus-community/elasticsearch_exporter) for
the collection of metrics from ElasticSearch servers.

> Note that currently, an Agent can only collect metrics from a single ElasticSearch server.
> However, the exporter is able to collect the metrics from all nodes through that server configured.

Reviewer: ```suggestion
{{% admonition type="note" %}}
Currently, an Agent can only collect metrics from a single ElasticSearch server.
However, the exporter can collect the metrics from all nodes through that server configured.
{{% /admonition %}}
```

We strongly recommend that you configure a separate user for the Agent, and give it only the strictly mandatory
security privileges necessary for monitoring your node, as per the [official documentation](https://github.com/prometheus-community/elasticsearch_exporter#elasticsearch-7x-security-privileges).

## Usage

```river
prometheus.exporter.elasticsearch "LABEL" {
    address = "ELASTICSEARCH_ADDRESS"
}
```

## Arguments

> The following arguments can be used to configure the exporter's behavior.

Reviewer: ```suggestion
You can use the following arguments to configure the exporter's behavior.
```

`targets` | `list(map(string))` | The targets that can be used to collect `elasticsearch` metrics.

For example, `targets` can either be passed to a `prometheus.relabel`
component to rewrite the metrics' label set, or to a `prometheus.scrape`
component that collects the exposed metrics.

The exported targets will use the configured [in-memory traffic][] address
specified by the [run command][].

[in-memory traffic]: {{< relref "../../concepts/component_controller.md#in-memory-traffic" >}}
[run command]: {{< relref "../cli/run.md" >}}

## Component health

`prometheus.exporter.elasticsearch` is only reported as unhealthy if given
an invalid configuration. In those cases, exported fields retain their last
healthy values.

## Debug information

`prometheus.exporter.elasticsearch` does not expose any component-specific
debug information.

## Debug metrics

`prometheus.exporter.elasticsearch` does not expose any component-specific
debug metrics.

## Example

This example uses a [`prometheus.scrape` component][scrape] to collect metrics
from `prometheus.exporter.elasticsearch`:

```river
prometheus.exporter.elasticsearch "example" {
  address = "localhost:9200"
}

> // Configure a prometheus.scrape component to collect Elasticsearch metrics.
> prometheus.scrape "demo" {
>   targets    = prometheus.exporter.elasticsearch.example.targets
>   forward_to = [ /* ... */ ]
> }

Reviewer 2: I think this should be as copy-paste-able as possible, so something like:
```suggestion
// Configure a prometheus.scrape component to collect Elasticsearch metrics.
prometheus.scrape "demo" {
  targets    = prometheus.exporter.elasticsearch.example.targets
  forward_to = [ prometheus.remote_write.example.receiver ]
}

prometheus.remote_write "example" {
  // Send metrics to a locally running Mimir.
  endpoint {
    url = "http://mimir:9090/api/v1/write"

    basic_auth {
      username = "sample-username"
      password = "sample-password"
    }
  }
}
```

Author: I haven't seen any example like this in the current docs. We could change it, but for now better to be consistent, no?

Reviewer 2: The guide on writing docs hasn't always been followed (and quite a few things predate it), but the guidance is to make it copy-pastable and a full pipeline, when possible: https://github.com/grafana/agent/blob/main/docs/developer/writing-flow-component-documentation.md?plain=1#L494

Reviewer 2: True, I'm not sure there's a stance on exactly _how_ copy-pastable the example needs to be, and it won't be possible to make it work out of the box for every case. But in this case it would be easy for someone to copy-paste the full example into a river config and would just need to replace the url.

It could also be made more generic with something like (which then wouldn't work if basic auth was needed):
```
prometheus.remote_write "example" {
  endpoint {
    url = env("REMOTE_WRITE_URL")
  }
}
```

Author: I think your point it's valid, but out of the scope of this PR IMHO. Do you want open issue/PR to address this?

Reviewer 2: Sure, sounds good to me :smile:
